support reference to board-provided dts include files from application overlays

pabigot opened the pull request:
The Particle mesh feather boards provide overlays that allow individual applications to select the SPI peripheral to be used for the pins associated with a specific labelled SPI device.  This is necessary because different peripheral instances have slightly different properties.

As currently implemented these files are not in the search path when generating the aggregate DTS file, so:
```
#include "mesh_feather_spi_spi1.dtsi"

&spi1 {
	/* add a peripheral here */
};
```
fails to work as `particle_xenon.overlay`.
    
The first commit adds the board directory to the search path when an overlay is being provided, so that overlays in application directories can include the desired DTSI file.

A second commit removes an inconsistency in dtsi files for Particle mesh feather devices.

pabigot committed:
cmake: DT: Add support for overlays that reference board-provided dtsi

The Particle mesh feather boards provide device-tree overlays that
allow individual applications to select the SPI peripheral to be
used for the pins associated with a specific labelled SPI device.
This is necessary because different peripheral instances have slightly
different properties.

Add BOARD_DIR to DTS_ROOTS so these shared files can be located when
included from application-specific overlays.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
boards: arm: particle_*: uncomment cs-gpios in spi assignments

At the time these overlays were created the author was unclear on how to
correctly represent SPI chip selects on a target where the peripheral
does not control that signal.  As such the pin assignment was present
only as documentation.

Enlightenment was subsequently achieved and the cs-gpios property should
be properly defined.

Note that the mesh feather does not define a chip select to be used with
the spi1 peripheral.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

SebastianBoe commented on cmake/dts.cmake:
```diff
@@ -49,6 +49,11 @@ if(SUPPORTS_DTS)
+    list(APPEND
+      DTS_ROOT_SYSTEM_INCLUDE_DIRS
+      -isystem ${BOARD_DIR}
+      )
```
Let us instead add the BOARD_DIR to the list of DTS_ROOT's.

Then it will be easier to document this behaviour, and we will need less CMake code.

Unfortunately . is not in the list of include paths for DTS_ROOTs, so any .dtsi file that needs to be included from an application overlay would need to first be first moved to BOARD_DIR/include, or BOARD_DIR/dts. But this is a small price to pay for system consistency IMHO.

pabigot replied:
This makes sense, but it's applying a standard to a new case: there are no board definitions that have `dts` or `include` subdirectories, but there are board definitions that have `dtsi` files in `.`.  This is probably the first time board-specific DTS options have been exercised.

Based on previous experience with #12377, which has some slight relation to this, I'd like confirmation/acknowledgement from @galak and @nashif that this approach is acceptable before moving things around.

pabigot replied:
Changes made.

pabigot committed:
boards: particle_*: move DTS includes to dts subdirectory

This conforms to the standard layout implied by adding BOARD_DIR to
DTS_ROOTS.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

SebastianBoe commented:
This needs to be documented, but unfortunately, it's not clear to me how or where.

I can see a definition of DTS_ROOT, but not any references to what the built-in DTS_ROOT's are. E.g. that it is possible for the application to define includable .dtsi files.

This is the relevant documentation:

https://docs.zephyrproject.org/latest/guides/dts/index.html#source-tree-hierarchy
https://docs.zephyrproject.org/latest/application/index.html#devicetree-definitions

Perhaps in the description of DTS_ROOT it should be stated what the default/built-in DTS_ROOT's are.

pabigot commented:
Documentation has:
```
    dts/bindings/
    dts/common/
    dts/${ARCH}/
    include/
```
`dts.cmake` has:
```
        include
        dts/common
        dts/${ARCH}
        dts
```

Is order preserved by the build infrastructure?  I would want application to take priority over board, and board over base; and the documentation should reflect the order that's actually used.

I'm not sure the proposed edit is correct.  `dts/bindings` may be important for application-specific YAML, but if it works that path must be being found through some other mechanism than `dts.cmake` adding it to an include.  So there are really two lists: the directories that must be present for general device-tree support, and the paths that are added for search by cmake (in their priority order).

The root `dts` should definitely be in the cmake paths.  Forcing board definition directories to add intervening arch directory under their `dts` subdirectory adds no value at all.

Finally there should be an explanation for why `include` is relevant to the DTS hierarchy and cmake, and clarification for how it's used in that context.  Device tree source and yaml files should **not** (IMO) be placed in `include`.

All this is a bit of scope creep, and not really my area.  I just pushed edits that I think provide the core description and leave the fiddly bits for somebody else.

SebastianBoe commented on doc/application/index.rst:
```diff
@@ -723,7 +723,17 @@ its location through the ``DTS_ROOT`` CMake Cache variable:
+    include
+    dts/common
+    dts/${ARCH}
+    dts
```
We should not have to repeat what happens when a directory is in the DTS_ROOT as this is already stated above.

> the other directories contain files that can be included from DT sources.

If you feel this is not being conveyed appropriately then we should update the original documentation instead of re-iterating it.

pabigot replied:
I've updated the documentation to use your proposed text, and changed the order and content to match what I think `dts.cmake` is doing.  Please verify.

I also updated the spelling of "Device Tree" in the nearby lines to match the most common spelling in that file.

SebastianBoe commented:
> Is order preserved by the build infrastructure? I would want
> application to take priority over board, and board over base and
> the documentation should reflect the order that's actually used.

Yes, the desired priority is matching the implemented priority.

> I'm not sure the proposed edit is correct. dts/bindings may be
> important for application-specific YAML, but if it works that path
> must be being found through some other mechanism than dts.cmake adding
> it to an include. So there are really two lists: the directories that
> must be present for general device-tree support, and the paths that
> are added for search by cmake (in their priority order).

The bindings directory, which contains yaml files as opposed to
DT files, is being added for each DTS_ROOT by dts.cmake as
intended.

This directory is used during post-processing after the DTC has
completed.

IMHO this is adequately described by this sentence (but any suggestions for
improvement are welcome):

"The binding directory contains bindings and the other
directories contain files that can be included from DT sources."

SebastianBoe reviewed (approved):
Great work, good to have this supported.

dbkinder commented on doc/application/index.rst:
```diff
@@ -698,16 +698,18 @@ Zephyr binary into your application directory.
 
-Additional DeviceTree directory trees, or DTS_ROOTs, can be added by
-creating this directory tree::
+Device Tree directory trees are found in``APPLICATION_SOURCE_DIR``,
```
```suggestion
Device Tree directory trees are found in ``APPLICATION_SOURCE_DIR``,
```

pabigot replied:
done

dbkinder reviewed (changes requested):
A couple of typos in the docs, but otherwise looks OK.

dbkinder commented on doc/application/index.rst:
```diff
@@ -698,16 +698,18 @@ Zephyr binary into your application directory.
 :file:`CMakeLists.txt` file.
 
-DeviceTree Definitions
+Device Tree Definitions
```
Actually, according to https://www.devicetree.org/ the correct spelling is ``devicetree`` (or ``Devicetree``, without a space.

dbkinder reviewed (changes requested):
https://www.devicetree.org/ says the spelling is without a space:  devicetree  (as it was previously written)

pabigot commented:
> https://www.devicetree.org/ says the spelling is without a space: devicetree (as it was previously written)

Yes; however I propose that changing this is not in scope for this PR as the change would touch at least eight different files under `doc/`, and possibly non-documentation files if we're going to be consistent.  I'd hoped to get away with slightly reduced inconsistency in the area being modified.  Will you accept that?

dbkinder reviewed (approved):
I wouldn't expect you to make the spelling of devicetree correct across all our documentation.  I just wanted you to put it back to the correct spelling for the changes you made.  I'll approve this PR and work on another that goes through all the documentation to make it consistent (that way a search will find all occurrences, for example).

pabigot commented:
OK, it's back to the original spelling.
